### ISSUE TITLE
Issue: Scheduling, Sampling, and Top-k Selection Risks in generate_dlm (Potential Instability & Performance Degradation)

### ISSUE THREAD

**Issue by red1239109-cmd** [furiosa-ai/ParallelBench#6]
While analyzing the core decoding loop in `generate_dlm`, several architectural and implementation-level issues were identified. These problems lead to runtime instability, incorrect performance measurements (NFE), degraded throughput, and potential crashes under specific configurations.

---

#### **1. Sampling Logic is Effectively Deterministic**

At the beginning of `generate_dlm`, the following assertions exist:

```python
assert temperature == 0
assert alg_temp == 0

```

This forces the `add_gumbel_noise` function to always return raw logits, effectively making the decoder operate in a fully deterministic `argmax` mode. This contradicts the documentation and implies that any sampling-based evaluation results may be misleading.

#### **2. Numerically Unstable & Non-standard Gumbel Formulation**

The current implementation of `add_gumbel_noise` deviates from the standard Gumbel-Max trick:

* **Current:** `logits.exp() / ((-torch.log(noise)) ** temperature)`
* **Standard:** , then 

**Risks:**

* **Overflow:** Using `logits.exp()` directly leads to high overflow risk.
* **Numerical Error:** `log(noise)` is undefined if `noise == 0`, leading to `NaN` or `Inf`.

#### **3. Performance Bottleneck: CPU-side Python Loops in Scheduling**

In `get_num_transfer_tokens_maskgit`, the scheduling ratio is calculated using nested Python loops:

```python
ratios = [[gamma_func((t+1) / steps, ...) for t in range(steps)] for total_num_item in total_num[:, 0]]

```

This approach causes significant overhead by forcing CPU-GPU synchronization and breaking tensor parallelism, especially as batch sizes or sequence lengths increase.

#### **4. Incorrect NFE (Number of Function Evaluations) Metric**

The `nfe` counter is initialized (`nfe = 0`) but never incremented within the forward pass loop. This renders any efficiency benchmarks or comparisons relying on NFE completely invalid.

#### **5. Out-of-Range Risk in `topk` Selection**

The code lacks a guard for cases where the requested  exceeds the number of available masked tokens:

```python
_, select_indices = torch.topk(confidence[j], k=num_transfer_tokens[j, i].item())

```

If the remaining masked tokens are fewer than `num_transfer_tokens`, the system will throw a `RuntimeError`.

#### **6. Frequent Device Synchronization (GPU → CPU Transfers)**

Inside the main decoding loop, multiple tensors are cloned and moved to the CPU in every iteration:

```python
intermediate_inputs.append(x.clone().cpu())
intermediate_results.append(x0.clone().cpu())

```

This forces constant device synchronization, breaking the GPU/NPU kernel pipeline and severely degrading throughput.

---

### **Suggested Fixes**

1. **Stochastic Sampling:** Remove the `assert temperature == 0` or implement a proper stochastic sampling path.
2. **Stable Gumbel-Max:** Replace the current formulation with the standard, numerically stable version:

3. **Vectorized Scheduling:** Replace Python loops in `gamma_func` logic with vectorized PyTorch operations on the GPU.
4. **Metric Correction:** Increment `nfe += 1` inside each forward pass iteration.
5. **Top-k Guard:** Add a safety guard: `k = min(k, int(mask_index[j].sum().item()))`.
6. **Optional History:** Move CPU-transfer history collection behind a `debug=True` flag to prevent production/benchmark slowdowns.

---

### **Impact**

These issues directly affect the reliability of performance benchmarks and the scalability of dLLM evaluations. Addressing these will ensure the benchmark's integrity and improve overall system stability.

---

감사합니다~!!
